File: app04_python_django_nft/backend/api/certificate_service.py
"""Orchestrates the artwork -> IPFS pinning -> on-chain minting -> Certificate persistence
pipeline. Mirrors the other ports' CertificateService.
"""

import datetime
import uuid

from django.core.exceptions import ValidationError

from api.errors import (
    ArtworkNotFoundException,
    ArtworkNotPinnedException,
    CertificateNotFoundException,
    DuplicateContentHashException,
    NotVerifiedException,
    RoyaltyOutOfRangeException,
)
from api.models import Artwork, ArtworkStatus, Certificate


def _now_iso8601():
    return datetime.datetime.now(datetime.timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")


class CertificateService:
    def __init__(self, storage, contract_service, kyc_service):
        self.storage = storage
        self.contract_service = contract_service
        self.kyc_service = kyc_service

    def create_artwork(self, upload, title, description, medium, year_created, royalty_percentage_bps,
                        artist_wallet_address, artist_did):
        if royalty_percentage_bps is None or royalty_percentage_bps < 0 or royalty_percentage_bps > 10000:
            raise RoyaltyOutOfRangeException(royalty_percentage_bps if royalty_percentage_bps is not None else -1)

        artwork = Artwork(
            id=uuid.uuid4(),
            file_id=uuid.uuid4(),
            original_filename=upload.original_filename,
            sha256_hash=upload.sha256_hash,
            title=title,
            description=description,
            medium=medium,
            year_created=year_created,
            royalty_percentage_bps=royalty_percentage_bps,
            artist_wallet_address=artist_wallet_address,
            artist_did=artist_did,
            status=ArtworkStatus.UPLOADED,
            created_at=_now_iso8601(),
        )

        image_pin = self.storage.pin_file(upload.content, upload.original_filename, upload.content_type)
        artwork.image_ipfs_uri = image_pin.uri

        metadata = {
            "title": artwork.title,
            "description": artwork.description,
            "medium": artwork.medium,
            "yearCreated": artwork.year_created,
            "artistDid": artwork.artist_did,
            "sha256Hash": artwork.sha256_hash,
            "image": image_pin.uri,
        }
        metadata_pin = self.storage.pin_json(metadata, f"{artwork.id}-metadata")
        artwork.metadata_ipfs_uri = metadata_pin.uri

        artwork.status = ArtworkStatus.PINNED
        artwork.save()
        return artwork

    def mint_certificate(self, artwork_id, recipient_address):
        try:
            artwork = Artwork.objects.get(id=artwork_id)
        except (Artwork.DoesNotExist, ValueError, ValidationError):
            raise ArtworkNotFoundException(artwork_id)

        if artwork.status != ArtworkStatus.PINNED:
            raise ArtworkNotPinnedException(artwork_id)
        if not self.kyc_service.is_verified(artwork.artist_wallet_address):
            raise NotVerifiedException(artwork.artist_wallet_address)
        if Certificate.objects.filter(content_hash_hex=artwork.sha256_hash).exists():
            raise DuplicateContentHashException(artwork.sha256_hash)

        mint_result = self.contract_service.mint_certificate(
            recipient_address, artwork.metadata_ipfs_uri, artwork.sha256_hash,
            artwork.artist_wallet_address, artwork.royalty_percentage_bps,
        )

        certificate = Certificate(
            token_id=mint_result.token_id,
            artwork=artwork,
            content_hash_hex=artwork.sha256_hash,
            contract_address=mint_result.contract_address,
            tx_hash=mint_result.tx_hash,
            owner_address=recipient_address,
            royalty_percentage_bps=artwork.royalty_percentage_bps,
            royalty_recipient=artwork.artist_wallet_address,
            minted_at=_now_iso8601(),
        )
        artwork.status = ArtworkStatus.MINTED
        artwork.save()
        certificate.save()
        return certificate

    def get_certificate(self, token_id):
        try:
            return Certificate.objects.select_related("artwork").get(token_id=token_id)
        except Certificate.DoesNotExist:
            raise CertificateNotFoundException(token_id)

    def get_certificates_for_wallet(self, wallet_address):
        return list(Certificate.objects.select_related("artwork").filter(owner_address=wallet_address))

File: app03_csharp_nft/backend/NftCerts/Certificates/CertificateService.cs
using System.Text.Json;
using NftCerts.Blockchain;
using NftCerts.Db;
using NftCerts.Errors;
using NftCerts.Identity;
using NftCerts.Storage;
using Microsoft.EntityFrameworkCore;

namespace NftCerts.Certificates;

// The hashed/watermarked bytes of an upload, retrievable by fileId when creating an artwork.
public record StoredUpload(string OriginalFilename, string ContentType, string Sha256Hash, byte[] Content);

// Orchestrates the artwork -> IPFS pinning -> on-chain minting -> Certificate persistence
// pipeline. Mirrors app01/app02's CertificateService.
public class CertificateService(
    NftCertsDbContext db,
    IIpfsStorageService ipfsStorageService,
    ContractService contractService,
    IKycVerificationService kycVerificationService)
{
    public Artwork CreateArtwork(StoredUpload upload, string title, string description, string medium,
                                  int? yearCreated, int? royaltyPercentageBps, string artistWalletAddress,
                                  string artistDid)
    {
        if (royaltyPercentageBps is null or < 0 or > 10000)
        {
            throw new RoyaltyOutOfRangeException(royaltyPercentageBps ?? -1);
        }

        var artwork = new Artwork
        {
            Id = Guid.NewGuid().ToString(),
            FileId = Guid.NewGuid().ToString(),
            OriginalFilename = upload.OriginalFilename,
            Sha256Hash = upload.Sha256Hash,
            Title = title,
            Description = description,
            Medium = medium,
            YearCreated = yearCreated,
            RoyaltyPercentageBps = royaltyPercentageBps.Value,
            ArtistWalletAddress = artistWalletAddress,
            ArtistDid = artistDid,
            Status = ArtworkStatus.UPLOADED,
            CreatedAt = NowIso8601(),
        };

        PinResult imagePin = ipfsStorageService.PinFile(upload.Content, upload.OriginalFilename, upload.ContentType);
        artwork.ImageIpfsUri = imagePin.Uri;

        var metadata = JsonSerializer.SerializeToElement(new
        {
            title = artwork.Title,
            description = artwork.Description,
            medium = artwork.Medium,
            yearCreated = artwork.YearCreated,
            artistDid = artwork.ArtistDid,
            sha256Hash = artwork.Sha256Hash,
            image = imagePin.Uri,
        });
        PinResult metadataPin = ipfsStorageService.PinJson(metadata, artwork.Id + "-metadata");
        artwork.MetadataIpfsUri = metadataPin.Uri;

        artwork.Status = ArtworkStatus.PINNED;

        db.Artworks.Add(artwork);
        db.SaveChanges();
        return artwork;
    }

    public Certificate MintCertificate(string artworkId, string recipientAddress)
    {
        Artwork artwork = db.Artworks.Find(artworkId) ?? throw new ArtworkNotFoundException(artworkId);

        if (artwork.Status != ArtworkStatus.PINNED)
        {
            throw new ArtworkNotPinnedException(artworkId);
        }
        if (!kycVerificationService.IsVerified(artwork.ArtistWalletAddress))
        {
            throw new NotVerifiedException(artwork.ArtistWalletAddress);
        }
        if (db.Certificates.Any(c => c.ContentHashHex == artwork.Sha256Hash))
        {
            throw new DuplicateContentHashException(artwork.Sha256Hash);
        }

        MintResult mintResult = contractService.MintCertificate(
            recipientAddress, artwork.MetadataIpfsUri, artwork.Sha256Hash, artwork.ArtistWalletAddress,
            artwork.RoyaltyPercentageBps);

        var certificate = new Certificate
        {
            TokenId = mintResult.TokenId,
            ArtworkId = artwork.Id,
            Artwork = artwork,
            ContentHashHex = artwork.Sha256Hash,
            ContractAddress = mintResult.ContractAddress,
            TxHash = mintResult.TxHash,
            OwnerAddress = recipientAddress,
            RoyaltyPercentageBps = artwork.RoyaltyPercentageBps,
            RoyaltyRecipient = artwork.ArtistWalletAddress,
            MintedAt = NowIso8601(),
        };

        artwork.Status = ArtworkStatus.MINTED;
        db.Certificates.Add(certificate);
        db.SaveChanges();
        return certificate;
    }

    public Certificate GetCertificate(long tokenId) =>
        db.Certificates.FirstOrDefault(c => c.TokenId == tokenId)
        ?? throw new CertificateNotFoundException(tokenId);

    public List<Certificate> GetCertificatesForWallet(string walletAddress) =>
        db.Certificates.Where(c => c.OwnerAddress == walletAddress).ToList();

    internal static string NowIso8601() => DateTime.UtcNow.ToString("yyyy-MM-dd'T'HH:mm:ss'Z'");
}
